### ISSUE TITLE
Allow json/BaseModels in query parameters

### ISSUE THREAD

**Issue by luckydonald** [tiangolo/fastapi#884]
### Is your feature request related to a problem

I have to support getting a structured json string as parameter:
<img width="891" alt="Screenshot 2020-01-17 at 22 08 55" src="https://user-images.githubusercontent.com/2737108/72646249-f86be980-3975-11ea-9e3c-3ea7816fd1e1.png">

```url
https://localhost:8080/test?moop={"name": "something %26 something else","description": "str" ,"price": 0.2,"tax": 0.2}&foo=bar
```

I would like to have a easy validation for those fields as well, as this could be automatic

### The solution you would like

```python
@routes.api_route('/test', methods=['GET', 'POST'], tags=['official', 'message', 'send'])
async def test(
    moop: Json[Item] = Query(..., description='Something something'),
    foo: str = Query(...),
):
    return {'woop': moop, 'lol': foo}
# end def
```

<details><summary>Initial version:</summary>

This was the initial thought, before I discovered the `Json` type.
```python
@routes.api_route('/test', methods=['GET', 'POST'], tags=['official', 'message', 'send'])
async def test(
    moop: Item = Query(..., description='Something something'),
    foo: str = Query(...),
):
    return {'woop': moop, 'lol': foo}
# end def
```

</details>

### Bug you face

```py
AssertionError: Param: moop can only be a request body, using Body(...)
```

### Describe alternatives you've considered

```py
    moop: str
```

**Comment by luckydonald** [comment]
Funny, using `moop: Json` is working:     

```py
from pydantic.types import Json

@routes.api_route('/test', methods=['GET', 'POST'], tags=['official', 'message', 'send'])
async def test(
    moop: Json = Query(..., description='Something something'),
    foo: str = Query(...),
):
  moop = Item(moop)
```

but  `moop: Json[Item]` is not working:

```py
@routes.api_route('/test', methods=['GET', 'POST'], tags=['official', 'message', 'send'])
async def test(
    moop: Json[Item] = Query(..., description='Something something'),
    foo: str = Query(...),
):
    moop = Item(moop)
    return {'woop': moop, 'lol': foo}
# end def
```
is failing with:
```py
File "/path/to/project/main.py", line 482, in <module>
  from .views.api.v4_5.sendable import routes as sendable_routes
File "/path/to/project/views/api/v4_5/sendable.py", line 85, in <module>
  foo: str = Query(...),
File "/path/to/3.7.venv/lib/python3.7/site-packages/fastapi/routing.py", line 487, in decorator
  callbacks=callbacks,
File "/path/to/3.7.venv/lib/python3.7/site-packages/fastapi/routing.py", line 431, in add_api_route
  callbacks=callbacks,
File "/path/to/3.7.venv/lib/python3.7/site-packages/fastapi/routing.py", line 334, in __init__
  self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
File "/path/to/3.7.venv/lib/python3.7/site-packages/fastapi/dependencies/utils.py", line 305, in get_dependant
  ), f"Param: {param_field.name} can only be a request body, using Body(...)"
AssertionError: Param: moop can only be a request body, using Body(...)
```

**Comment by dmontagu** [comment]
I think what's happening here is that FastAPI tries to only allow certain types to be parsed automatically from query parameters (presumably to limit confusion where you are expecting it to parse that data from the *body*, rather than the query parameters; obviously not relevant in *this* case).

When you pass `Json`, the implicit generic parameter is treated as `Any`, which is apparently being allowed. (My guess is this is somehow related to the fact that `Json[Any]` could be parsing a `str`, for example, which is allowed as a query parameter by FastAPI.)

However, when you say `Json[Item]`, you are now making it clear you want to parse a dict out of the query parameters, but fastapi explicitly only allows parsing primitives or sequences of primitives, so that's why you are getting this error.

It's a little less ergonomic, but I would recommend creating a custom dependency that parses the query string into the type you want. Something like:

```python
def get_moop(moop: Json[Any] = Query(..., description='Something something')) -> Item:
    return Item.parse_obj(moop)

async def test(
    moop: Item = Depends(get_moop),
    foo: str = Query(...),
):
```

------

From what I gather, the [OpenAPI spec](https://swagger.io/docs/specification/describing-parameters/#query-parameters) *does* allow you to pass objects as query parameter values, so in theory I don't see why this shouldn't be supported. But as far as I can tell, it isn't supported right now.

I'd be happy to review a PR if someone cared to put in the work to implement this. (Note: you might have to manually parse the raw query string, since I don't think starlette does that automatically.)

**Comment by luckydonald** [comment]
@dmontagu  
Interestingly, using `Json` as the type already converts the argument to a valid python objects, in this case a dict.

So basically all whats left is throwing that json already parsed as python primitives into the normal object loader as specified in the square brackets.

However, I don't really understand the current code which is raising an error, what I have to change there, and I also don't know what's the best place to add that transformation.

**Comment by luckydonald** [comment]
@dmontagu Going back to the original post I now edited:
I'm not sure which is better,
- `variable: Json[SomeType] = Query(...)`
- `variable: SomeType = Query(...)`.

I believe the first step is to recognize that json should be allowed as `Query`.
I don't know if that `Json` type is for cases in the body where there would be an actual json encoded string.
So the `"real_json"` key in the following example:
```json
{
    "real_json": "{\"sample_text\": \"foobar\"}",
    "not_json": {"sample_text": "foobar"}
}
```
I wonder how extendable that is, because I imagine it could also contain XML or some other format as well. (I dunnow, can we parse XML by chance, or have other json-like formats?)

As starlette correctly parses `field: Json` to be a python dict/list/int/float/None object, I'd actually rather go with `Json`.

**Comment by luckydonald** [comment]
With the previous comment in mind I see the following cases:

- `variable: SomeType`: Use the json body, and load the `"variable": {...}` part as `SomeType`.
- `variable: Json[SomeType]`: Use the json body, and decode the json encoded`"variable": "{...}"` and  then load the part as `SomeType`.
- `variable: SomeType = Query(...)`: 
  - Should that be indeed an error?
  - Like maybe `AssertionError: Param: variable containing a model can only be a part of a request body, using the value Body(...) or of type Json or Json[...] using the value Query(...) or Path(...)`
- `variable: Json[SomeType] = Query(...)`: We specified we wanna use the query parameter as Json, and thus we do the same lo
- `variable: Json = Query(...)`: We specified we wanna use the query parameter as Json, but no model, so we keep parsing the json to python native types. This is how it already works.
- `variable: Json[SomeType] = Query(...)`: We specified we wanna use the query parameter as Json, and because we specified some type, that is validated and loaded in the typical Pydantic way.
- `variable: Json = Path(...)`: See `variable: Json = Query(...)`.
- `variable: Json[SomeType] = Path(...)`: See `variable: Json[SomeType] = Query(...)`.

**Comment by luckydonald** [comment]
Okey, I made a first implementation attempt.

# Test cases
```py
from pydantic import BaseModel

class TestModel(BaseModel):
    name: str
    description: str
    price: float
    tax: float
# end class
```


# case 1, `Query`
```py
@routes.api_route('/test1', methods=['GET', 'POST'], tags=['official', 'message', 'send'])
async def test1(
    moop: Json[TestModel] = Query(..., description='Something something'),
    foo: str = Query(...),
):
    return {'woop': moop, 'lol': foo}
# end def
```
```curl
curl --location --request GET 'http://127.0.0.1:8080/test1?foo=test123&moop={%22name%22:%20%22test%22,%20%22description%22:%20%22sample%20text%22,%20%22price%22:%204458.69,%20%22tax%22:%2042.0}'
```
```json
{
    "woop": {
        "name": "test",
        "description": "sample text",
        "price": 4458.69,
        "tax": 42.0
    },
    "lol": "test123"
}
```
# case 2, `Path`
```py
@routes.api_route('/test2/{moop}', methods=['GET', 'POST'], tags=['official', 'message', 'send'])
async def test2(
    moop: Json[TestModel] = Path(..., description='Something something'),
    foo: str = Query(...),
):
    return {'woop': moop, 'lol': foo}
# end def
```
```curl
curl --location --request GET 'http://127.0.0.1:8080/test2/{"name": "test", "description": "sample text", "price": 4458.69, "tax": 42.0}?foo=test123'
```
```json
{
    "woop": {
        "name": "test",
        "description": "sample text",
        "price": 4458.69,
        "tax": 42.0
    },
    "lol": "test123"
}
```

# case 3, `Header`
```py
@routes.api_route('/test3', methods=['GET', 'POST'], tags=['official', 'message', 'send'])
async def test3(
    moop: Json[TestModel] = Header(..., description='Something something'),
    foo: str = Query(...),
):
    return {'woop': moop, 'lol': foo}
# end def
```
```curl
curl --location --request GET 'http://127.0.0.1:8080/test3?foo=test123' \
--header 'moop: {"name": "test", "description": "sample text", "price": 4458.69, "tax": 42.0}'
```
```json
{
    "woop": {
        "name": "test",
        "description": "sample text",
        "price": 4458.69,
        "tax": 42.0
    },
    "lol": "test123"
}
```

**Comment by luckydonald** [comment]
Okey, I prepared PR #908 for feedback.

**Comment by tiangolo** [comment]
Thanks for all the discussion here @luckydonald and @dmontagu !

As this is more or less a corner case that wouldn't necessarily be documented with OpenAPI I think the best way to solve it would be with a dependency as @dmontagu suggests rather than modifying FastAPI itself.

**Comment by diegopy** [comment]
Hi @tiangolo, regarding your last comment, this case is indeed documented by OpenAPI [here](https://swagger.io/docs/specification/describing-parameters/). Parameters can be in the path, query, header and cookie. In all 4 cases they can be encoded as `schema` or `content`, as described in the **schema vs content** section of that doc page. `content` in this case casn use json to encode a complex object in any of those 4 parameter types. Swagger UI even renders it properly. I made a sample schema (attached) as an example. There's also a screenshot of how Swagger UI handles this case perfect.
In summary, I would kindly ask if you could please reconsider adding FastAPI support for this fully documented and conformant use case. I volunteer to help in implementing it with your guidance.
Thank you.

![Screenshot_2020-03-31 Swagger Editor](https://user-images.githubusercontent.com/2140622/77989933-029f5200-72ee-11ea-9a11-c0dfca566003.png)
[openapi-query-arg.json.txt](https://github.com/tiangolo/fastapi/files/4406883/openapi-query-arg.json.txt)

**Comment by luckydonald** [comment]
@tiangolo I think the bot having this automatically closed is wrong, please reopen.

**Comment by tiangolo** [comment]
@diegopy , that's right, it is indeed possible to define several encoding styles in OpenAPI.

Now, here's the thing. The request parsing and extraction of data (from query, headers, cookies, etc) is done in Starlette, not in FastAPI. Implementing all the possible styles available in OpenAPI would not only mean possibly diverging from Starlette, but also adding an extreme extra maintenance burden to cover all the corner cases. So it's possible that some first steps have to be done as Pull Requests to Starlette.

About `content`, it's here in the spec: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#parameter-object

And it says:

> For more complex scenarios, the content property can define the media type and schema of the parameter.

As it is intended for more complex scenarios and it's not the default in the spec, it wouldn't make sense to make it a default.

So, implementing it would mean, after checking that the standard parsing from Starlette works, checking if there's a Pydantic model in a query parameter, making sure that it's an explicit `Query()` (as the default would be a body), and checking for lists of Pydantic models, etc. But making sure that those are lists of Pydantic models, as lists of scalars are supported normally.

I guess the idea would be not to support any of the extra `style`s, but only the `content` with JSON for the most complex types.

If you feel like taking a stab at it, that would be great, we could start with the simplest use case and move from there.

The simplest idea I could imagine would be adding a parameter `json=False` to `Query()`, `Header()`, etc. And trying to parse as JSON only when it is explicitly set to `True`.

**Comment by luckydonald** [comment]
@tiangolo could you reopen this issue then, please?

**Comment by ToGoBananas** [comment]
```
async def get_events_risks(
    request: Request,
    filter: EventsRisksFilterRequest = Query(EventsRisksFilterRequest(event_status=BetEventStatusEnum.PRE_LIVE)),
    db: Database = Depends(deps.get_db),
    user: User = Depends(deps.get_admin_user),
):
    filter = EventsRisksFilterRequest(**request.query_params)
    return await EventsRisksAdminManager().list(request_filter=filter)
```

Here is dirty hack how to achive desired behaviour

**Comment by w4tsn** [comment]
I'd like to ask again if this issue can be re-opened. I'm currently implementing a analytics backend which requires some rather complex query parameters in GET requests. I really enjoy writing models and schemas once and use them all out the code to just relax and watch fastapi to build the API documentation for me. It's a real bummer that dirty hacks and dependency-work arounds (including me writing OpenAPI spec manually) are necessary to get this to work for every new function I implement.

**Comment by luckydonald** [comment]
Yes, as my original use case is still valid, I vote for that as well, @tiangolo.

**Comment by csudcy** [comment]
Just started using FastAPI & I'm surprised to find out FastAPI can't parse a JSON encoded query parameter into a Pydantic model & keep the docs good (cause so far it seems to be amazing for everything else).

FYI @ToGoBananas, I tried your fork (cause https://github.com/ToGoBananas/fastapi/pull/1/files makes it look like a very easy change) but, as far as I could tell, the docs still show the payload as the body of the request (which Chrome blocks for GET requests).

POST request it is then 😅

**Comment by ToGoBananas** [comment]
@csudcy 
Actually, it works quite well and generates proper open API spec.

![image](https://user-images.githubusercontent.com/5418148/121787449-c1cbea00-cbce-11eb-99d9-b8a0ca2a2139.png)

**Comment by AndreiPashkin** [comment]
@tiangolo 
> As this is more or less a corner case that wouldn't necessarily be documented with OpenAPI I think the best way to solve it would be with a dependency as @dmontagu suggests rather than modifying FastAPI itself.

OpenAPI 3.0 actually supports this use case. 
References:
https://github.com/OAI/OpenAPI-Specification/issues/552#issuecomment-281837187
https://swagger.io/specification/#parameter-object-examples

Example encoding from OpenAPI specs:
```
{
  "in": "query",
  "name": "coordinates",
  "content": {
    "application/json": {
      "schema": {
        "type": "object",
        "required": [
          "lat",
          "long"
        ],
        "properties": {
          "lat": {
            "type": "number"
          },
          "long": {
            "type": "number"
          }
        }
      }
    }
  }
}
```

AFAIK currently FastAPI doesn't support this but I think it should since OpenAPI does support it.

**Comment by justinttl** [comment]
@tiangolo 

Is there a way to define 

```
  "content": {
    "application/json": {
```

at the root of the parameter object with `Query`? It seems that any extra kwargs passed in goes straight into `schema` which does not allow for the nice swagger rendering that @diegopy posted above. 

For context, my team is able to define a pretty solid dependency based utility for doing custom JSON query args serde, but the swagger page leaves much to be desired. Having the ability to define top level openapi parameter config will be the last hurdle for us.

**Comment by andrescevp** [comment]
Follwing this thread... I have a workaround.

query parameter definition
```
_filters: Json =  Query(
            _filters_defult,
            parameter_openapi=json.loads(SqlFilters.schema_json(ref_template='#/components/schemas'))),
```


In the setup

```
extra_model_schemas = {
        "SqlFilters": json.loads(SqlFilters.schema_json(ref_template='#/components/schemas')),
    }

def custom_openapi(app: FastAPI):
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
                title=app.title,
                version=app.version,
                openapi_version=app.openapi_version,
                description=app.description,
                terms_of_service=app.terms_of_service,
                contact=app.contact,
                license_info=app.license_info,
                routes=app.routes,
                tags=app.openapi_tags,
                servers=app.servers,
            )

    if "components" not in openapi_schema.keys():
        openapi_schema["components"] = {}
    openapi_schema["components"]["schemas"] = {**openapi_schema["components"]["schemas"], **extra_model_schemas}
    app.openapi_schema = openapi_schema
    return app.openapi_schema


app.openapi = lambda: custom_openapi(app)

```

**Comment by imankulov** [comment]
With some magic of [python-makefun](https://github.com/smarie/python-makefun) from @smarie, I created a good enough generic solution for parsing models in query strings.

Usage example:

```python
class User(BaseModel):
    name: str

@app.get("/")
def root(user: User = json_param("user", User, description="User object")):
    return {"message": f"Hello, {user!r}"}
```

The code of `json_param()` is available in the gist snippet by the link.

https://gist.github.com/imankulov/cef71dd5a01f9a27caeb66f7bedaf241

**Comment by luckydonald** [comment]
@tiangolo  as you're going through the tickets, can you reopen this? It is still not really solved.

**Comment by atrigent** [comment]
So, let me get this straight. Not only does FastAPI by default assume that model-typed parameters on GET route functions are passed in the request body, despite request bodies on GET requests being strongly discouraged, it also actively _gets in my way_ when I try to correct that obviously incorrect default?
